### PR TITLE
7903781: Report the process id of the agent or other VM that was used for a jtreg action

### PR DESCRIPTION
**Pull request by jaikiran**
Can I please get a review of this change which proposes to repor the process id of the process that was used to run a jtreg action? This change addresses https://bugs.openjdk.org/browse/CODETOOLS-7903781.

As noted in that linked enhancement request, this change will report the process id  the `section` messages of each action. This will now look like:

```
#section:compile
----------messages:(7/297)----------
command: compile test/streams/FileDescriptorTest.java
reason: .class file out of date or does not exist
started: Fri Jul 26 00:56:01 UTC 2024
Mode: othervm
Process id: 88469
finished: Fri Jul 26 00:56:01 UTC 2024
elapsed time (seconds): 0.254

....

#section:main
----------messages:(7/232)----------
command: main FileDescriptorTest
reason: User specified action: run main FileDescriptorTest 
started: Fri Jul 26 00:56:01 UTC 2024
Mode: othervm
Process id: 88470
finished: Fri Jul 26 00:56:01 UTC 2024
elapsed time (seconds): 0.042
```
Notice the new "Process id: ..." line in those sections.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [CODETOOLS-7903781](https://bugs.openjdk.org/browse/CODETOOLS-7903781): Report the process id of the agent or other VM that was used for a jtreg action (**Enhancement** - P4)

### Reviewers
 * [Jonathan Gibbons](https://openjdk.org/census#jjg) (@jonathan-gibbons - **Reviewer**)
 * [Christian Stein](https://openjdk.org/census#cstein) (@sormuras - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jtreg.git pull/215/head:pull/215` \
`$ git checkout pull/215`

Update a local copy of the PR: \
`$ git checkout pull/215` \
`$ git pull https://git.openjdk.org/jtreg.git pull/215/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 215`

View PR using the GUI difftool: \
`$ git pr show -t 215`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jtreg/pull/215.diff">https://git.openjdk.org/jtreg/pull/215.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jtreg/pull/215#issuecomment-2252035339)

### CODE REVIEW

File: src/share/classes/com/sun/javatest/regtest/exec/Agent.java [hunk 1]
```diff
@@ -593,6 +576,17 @@ public int getId() {
         return id;
     }
 
+    /**
+     * Returns the process id of the {@code AgentServer} with which this {@code Agent}
+     * communicates. Returns {@code -1} if the process id of the {@code AgentServer}
```
Comment (jonathan-gibbons) [inline src/share/classes/com/sun/javatest/regtest/exec/Agent.java:581]: As a matter of style, it is better if the first sentence is "complete", so I suggest changing ". Returns" to " or ".

```
      * Returns the process id of the {@code AgentServer} with which this {@code Agent}
      * communicates or {@code -1} if the process id of the {@code AgentServer}
      * couldn't be determined.
```

File: src/share/classes/com/sun/javatest/regtest/util/ProcessUtils.java [hunk 2]
```diff
@@ -27,20 +27,35 @@
...
         }
+        destroyForciblyMethod = destroyMethod;
+
+        Method pidMethod = null;
+        try {
+            pidMethod = Process.class.getDeclaredMethod("pid"); // only available in Java 9+
+        } catch (NoSuchMethodException e) {
+            pidMethod = null;
+        }
+        PID_METHOD = pidMethod;
```
Comment (jonathan-gibbons) [inline src/share/classes/com/sun/javatest/regtest/util/ProcessUtils.java:58]: These two blocks of code are functionally similar but stylistically different in the names used.
I suggest making them similar, perhaps changing to use DESTROY_FORCIBLY_METHOD.

**Review by jonathan-gibbons — COMMENTED** [review]
Mostly good; some minor comments/suggestions.

File: src/share/classes/com/sun/javatest/regtest/util/ProcessUtils.java [hunk 3]
```diff
@@ -27,20 +27,35 @@
...
+import java.util.Objects;
 
 /**
  * Utilities for handling Processes.
  */
 public class ProcessUtils {
 
-    private static Method destroyForciblyMethod;
+    private static final Method destroyForciblyMethod;
+    private static final Method PID_METHOD;
```
Comment (jonathan-gibbons) [inline src/share/classes/com/sun/javatest/regtest/util/ProcessUtils.java:38]: How much we we still need reflective access to these methods?  
If they are only invoked from the `jtreg` tool (and never from a test VM), then the minimum level of JDK required to run `jtreg` is currently 11, right? And, we no longer support running test-VM code on anything less than 8, right?

So, at some level, it seems we may not need `ProcessUtils` any more -- unless maybe as a convenience wrapper.

File: src/share/classes/com/sun/javatest/regtest/exec/Agent.java (hunk 1 above)
Comment (jaikiran) [reply src/share/classes/com/sun/javatest/regtest/exec/Agent.java]: Done. I've updated the PR to use your suggested text.

File: src/share/classes/com/sun/javatest/regtest/util/ProcessUtils.java (hunk 2 above)
Comment (jaikiran) [reply src/share/classes/com/sun/javatest/regtest/util/ProcessUtils.java]: That makes sense - I've renamed that field to this suggested name.

File: src/share/classes/com/sun/javatest/regtest/util/ProcessUtils.java (hunk 3 above)
Comment (jaikiran) [reply src/share/classes/com/sun/javatest/regtest/util/ProcessUtils.java]: Hello Jon,

> And, we no longer support running test-VM code on anything less than 8, right?

I think you are right. Although, the `Process.pid()` is only available in Java 9+, I think that call is only run on the `jtreg` tool's VM which as you note seems to require Java 11 (I will have to verify some JDK CI runs to be sure that's the case currently). 

If it's OK, then I'll take up the usage and possible remove of this reflection code as a separate issue and verify things work fine without it.

Comment (sormuras) [reply src/share/classes/com/sun/javatest/regtest/util/ProcessUtils.java]: I created [CODETOOLS-7903789](https://bugs.openjdk.org/browse/CODETOOLS-7903789) follow-up issue to replace those reflection-based method invocations. Let's keep the scope of changes of the PR small(er), in order to release jtreg 7.5 soon(er).

**Review by sormuras — APPROVED** [review]

**Review by jonathan-gibbons — APPROVED** [review]

**Comment by jaikiran** [comment]
Thank you Jon and Christian for the reviews.

**Pull request closed without merging**